src/hotspot/cpu/x86/x86.ad

```diff
@@ -8615,56 +8629,85 @@ instruct vmuladdaddS2I_reg(vec dst, vec src1, vec src2) %{
+    BasicType bt = Matcher::vector_element_basic_type(this, $src);
+    __ vector_popcount_integral_evex(bt, $dst$$XMMRegister, $src$$XMMRegister, k0, true, vlen_enc);
+    // TODO: Once auto-vectorizer supports ConvL2I operation, PopCountVL
+    // should be succeeded by its corresponding vector IR and following
+    // special handling should be removed.
+    if (opcode == Op_PopCountVL && Matcher::vector_element_basic_type(this) == T_INT) {
+      __ evpmovqd($dst$$XMMRegister, $dst$$XMMRegister, vlen_enc);
+    }
   %}
   ins_pipe( pipe_slow );
 %}
 
-instruct vpopcountI(vec dst, vec src, vec xtmp1, vec xtmp2, vec xtmp3, rRegP rtmp, rFlagsReg cc) %{
-  predicate(!VM_Version::supports_avx512_vpopcntdq());
-  match(Set dst (PopCountVI src));
-  effect(TEMP dst, TEMP xtmp1, TEMP xtmp2, TEMP xtmp3, TEMP rtmp, KILL cc);
-  format %{ "vector_popcount_int  $dst, $src\t! using $xtmp1, $xtmp2, $xtmp3, and $rtmp as TEMP" %}
+instruct vpopcount_integral_reg_evex_masked(vec dst, vec src, kReg mask) %{
+  predicate((((Matcher::vector_element_basic_type(n->in(1)) == T_INT ||
+               Matcher::vector_element_basic_type(n->in(1)) == T_LONG)) &&
+             VM_Version::supports_avx512_vpopcntdq()) ||
+            (is_subword_type(Matcher::vector_element_basic_type(n->in(1))) &&
+             VM_Version::supports_avx512_bitalg()));
+  match(Set dst (PopCountVI src mask));
+  match(Set dst (PopCountVL src mask));
+  format %{ "vector_popcount_integral_evex_masked $dst, $src, $mask" %}
   ins_encode %{
     assert(UsePopCountInstruction, "not enabled");
-    int vlen_enc = vector_length_encoding(this);
-    __ vector_popcount_int($dst$$XMMRegister, $src$$XMMRegister, $xtmp1$$XMMRegister, $xtmp2$$XMMRegister,
-                           $xtmp3$$XMMRegister, $rtmp$$Register, vlen_enc);
+    int opcode = this->ideal_Opcode();
+    int vlen_enc = vector_length_encoding(this, $src);
+    BasicType bt = Matcher::vector_element_basic_type(this, $src);
+    __ evmovdquq($dst$$XMMRegister, $src$$XMMRegister, vlen_enc);
+    __ vector_popcount_integral_evex(bt, $dst$$XMMRegister, $src$$XMMRegister, $mask$$KRegister, true, vlen_enc);
   %}
   ins_pipe( pipe_slow );
 %}
 
-instruct vpopcountL_popcntd(vec dst, vec src) %{
-  predicate(VM_Version::supports_avx512_vpopcntdq());
-  match(Set dst (PopCountVL src));
-  format %{ "vector_popcount_long  $dst, $src\t! vector popcount packedL" %}
+instruct vpopcountI_avx_reg(vec dst, vec src, vec xtmp1, vec xtmp2, rRegP rtmp) %{
+  predicate((!VM_Version::supports_avx512_vpopcntdq() && Matcher::vector_element_basic_type(n->in(1)) == T_INT) ||
+            (!VM_Version::supports_avx512_bitalg() && is_subword_type(Matcher::vector_element_basic_type(n->in(1)))));
+  match(Set dst (PopCountVI src));
+  effect(TEMP dst, TEMP xtmp1, TEMP xtmp2, TEMP rtmp);
+  format %{ "vector_popcount_int  $dst, $src\t! using $xtmp1, $xtmp2 and $rtmp as TEMP" %}
   ins_encode %{
     assert(UsePopCountInstruction, "not enabled");
```

Comment:
The assert needs to be removed on this path.

Reply:
Same a above, it should be OK to keep this as is since its already used in match_rule_supported method.